src/server/pkg/storage/fileset/fileset.go

```diff
@@ -0,0 +1,87 @@
+package fileset
+
+import (
+	"bytes"
+	"io"
+
+	"github.com/pachyderm/pachyderm/src/server/pkg/storage/obj"
+)
+
+// Notes:
+// - I am drawing a distinction between parts of a fileset and the merge log because I think it makes sense to require
+// the full merging of data within a commit. The merge log will apply across commit boundaries. This also makes the finishing
+// step a bit easier because we simply merge everything in memory and in the scratch space, then maybe merge with the parent commits
+// based on some heuristics.
+// - This should be used primarily for a fileset that requires Put/Overwrite/Delete operations (input repos, uploading datum/chunk data).
+// For merging multiple serialized filesets, a Scratch and/or Merge operation should be used.
+// - Example structure in object storage:
+// /pfs/repo/commit/shard-0/fileset
+// /pfs/repo/commit/shard-0/index
+// /pfs/repo/commit/shard-0/mergelog
+// /pfs/repo/commit/shard-1/fileset
+// /pfs/repo/commit/shard-1/index
+// /pfs/repo/commit/shard-1/mergelog
+// /pfs/chunks/8cfd
+// /pfs/chunks/ef7d
+// /pfs/chunks/823a
+// /pfs/chunks/b54d
+// - Handling of shards will happen outside of the storage layer (does not make sense for storage layer to understand
+// workers and how to filter chunk data from them.
```

Comment:
I'm not totally sure what "parts of a fileset" and "the merge log" mean. but the splitting of these various ideas and introduction of a bunch of new nouns that seem to have a lot of overlapping functionality seems like a big weakness of this design. I don't think we're going to be able to have a design that's long term maintainable if we fragment concepts this much early on.

I also don't think it's going to work to have a bunch of different semantic files for each commit such as `shard-0/fileset`, `shard-1/index`. Doing that means that we can no longer overwrite a commits object with a coalesced version, because we could crash after updating one file but not another. Having shards explicitly defined in the object structure also seems like it's going to lead to some difficulties in long term maintainability and also lead to problems like the one we have now where you can't change the sharding of a pipeline after it's run.

Reply:
Yeah, I am a little unsure about the parts vs merge log. What I was thinking is that parts are things that show up within commit creation and are resolved at commit finish and the merge log would be relevant across commits. My thought was we would not create chunks for parts and instead only chunk the merged commit. Creating a large fileset would involve using the Scratch space and generating parts that would be collapsed at commit finish. This could be unified into the merge log and we could always chunk anything that leaves memory, but that makes taking advantage of disk a bit trickier, I think, and leaves the door open to creating a lot of garbage (in terms of chunks) that needs to be cleaned up, which may be fine.

Reply:
Having a separate prefix for each shard is probably not the right choice, but I don't think we are ever going to want to overwrite a commit's object with a coalesced version. I think what we want is a separate fileset that is coalesced at that commit/HEAD. Leaving around the new data from the commit allows us to essentially keep a diff if we want to cleanup the coalesced version. Also, the objects in the current design would be fileset, mergelog, and checkpoint.